internal/deb/extract.go

```diff
@@ -277,6 +282,140 @@ func extractData(dataReader io.Reader, options *ExtractOptions) error {
 	return nil
 }
 
+type pendingHardLink struct {
+	path         string
+	extractInfos []ExtractInfo
+}
+
+type extractHardLinkOptions struct {
+	*ExtractOptions
+	pendingLinks map[string][]pendingHardLink
+}
+
+// extractHardLinks iterates through the tarball a second time to extract the
+// hard links that were not extracted in the first pass.
+func extractHardLinks(pkgReader io.ReadSeeker, opts *extractHardLinkOptions) error {
+	offset, err := pkgReader.Seek(0, io.SeekStart)
+	if err != nil {
+		return err
+	}
+	if offset != 0 {
+		return fmt.Errorf("internal error: cannot seek to the beginning of the package")
+	}
+	dataReader, err := getDataReader(pkgReader)
+	if err != nil {
+		return err
+	}
+	defer dataReader.Close()
+
+	tarReader := tar.NewReader(dataReader)
+	for {
+		tarHeader, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		sourcePath := sanitizeTarPath(tarHeader.Name)
+		if sourcePath == "" {
+			continue
+		}
+
+		links, ok := opts.pendingLinks[sourcePath]
+		if !ok || len(links) == 0 {
+			continue
+		}
+
+		// For a target path, the first hard link will be created as a file with
+		// the content of the target path. If there are more pending hard links,
+		// the remaining ones will be created as hard links with the newly
+		// created file as their target.
+		absLink := filepath.Join(opts.TargetDir, links[0].path)
+		// Extract the content to the first hard link path.
+		createOptions := &fsutil.CreateOptions{
+			Path: absLink,
+			Mode: tarHeader.FileInfo().Mode(),
+			Data: tarReader,
+		}
+		err = opts.Create(links[0].extractInfos, createOptions)
+		if err != nil {
+			return err
+		}
+
+		// Create the remaining hard links.
+		for _, link := range links[1:] {
+			createOptions := &fsutil.CreateOptions{
+				Path: filepath.Join(opts.TargetDir, link.path),
+				Mode: tarHeader.FileInfo().Mode(),
+				// Link to the first file extracted for the hard links.
+				Link: absLink,
+			}
+			err := opts.Create(link.extractInfos, createOptions)
+			if err != nil {
+				return err
+			}
+		}
+		delete(opts.pendingLinks, sourcePath)
+	}
+
+	// If there are pending links, that means the link targets do not come from
+	// this package.
+	if len(opts.pendingLinks) > 0 {
+		var errs []string
+		for target, links := range opts.pendingLinks {
+			for _, link := range links {
+				errs = append(errs, fmt.Sprintf("cannot create hard link %s: no content at %s",
+					link.path, target))
+			}
+		}
+		if len(errs) == 1 {
+			return fmt.Errorf("%s", errs[0])
+		}
+		sort.Strings(errs)
+		return fmt.Errorf("\n- %s", strings.Join(errs, "\n- "))
```

Comment:
Looks like that will end up quite unusual, without a prefix, and with long repeated messages.

This is an unusual case, so I suggest simply stopping at the first problem and returning the error.

Reply:
Agree, that was in fact one of the options we discussed precisely for the same reason. Done in https://github.com/canonical/chisel/pull/179/commits/7616cb743bd38cb045dfdc3c1dbc6c1dc7cd9c29.